server/src/main/java/org/candlepin/model/EntitlementCurator.java

```diff
@@ -573,60 +573,69 @@ private boolean poolOverlapsRange(Pool p, Date startDate, Date endDate) {
      * input entitlements.
      * When dealing with large amount of entitlements for which it is necessary
      * to determine their modifier products.
-     * @param entitlement
-     * @return Entitlements that are being modified by the input entitlements
+     * @param entitlements
+     * @return number of Entitlements that are being modified by the input entitlements
      */
-    public Collection<String> batchListModifying(Iterable<Entitlement> entitlements) {
-        List<String> eids = new LinkedList<String>();
+    public int markModifyingEntsDirty(Iterable<Entitlement> entitlements) {
+        int result = 0;
 
         if (entitlements != null && entitlements.iterator().hasNext()) {
-            String hql =
-                "SELECT DISTINCT eOut.id" +
-                "    FROM Entitlement eOut" +
-                "        JOIN eOut.pool outPool" +
-                "        JOIN outPool.providedProducts outProvided" +
-                "        JOIN outProvided.productContent outProvContent" +
-                "        JOIN outProvContent.content outContent" +
-                "        JOIN outContent.modifiedProductIds outModProdId" +
-                "    WHERE" +
-                "        outPool.endDate >= current_date AND" +
-                "        eOut.owner = :owner AND" +
-                "        eOut NOT IN (:ein) AND" +
-                "        EXISTS (" +
-                "            SELECT eIn" +
-                "                FROM Entitlement eIn" +
-                "                    JOIN eIn.consumer inConsumer" +
-                "                    JOIN eIn.pool inPool" +
-                "                    JOIN inPool.product inMktProd" +
-                "                    LEFT JOIN inPool.providedProducts inProvidedProd" +
-                "                WHERE eIn in (:ein) AND inConsumer = eOut.consumer AND" +
-                "                    inPool.endDate >= outPool.startDate AND" +
-                "                    inPool.startDate <= outPool.endDate AND" +
-                "                    (inProvidedProd.id = outModProdId OR inMktProd.id = outModProdId)" +
-                "        )";
 
-            Query query = this.getEntityManager().createQuery(hql);
+            String sql = "UPDATE cp_entitlement" +
+                "     SET dirty = true" +
+                "     WHERE id IN (" +
```

Comment:
I am not sure I see the problem, I thought the limit clause should fix it, but perhaps I dont see it yet. could you describe a problem scenario please? or better yet, demonstrate the issue on a db? let us chat online.